### PR TITLE
Fix IOException when server responds with bad request.

### PR DESCRIPTION
**Pull request by stefan-dimitrov**
Fix for issue: #12 and #5

### CODE REVIEW

File: src/main/java/com/clouway/chita/ChitaHttpClient.java [hunk 1]
```diff
@@ -95,7 +95,11 @@ public HttpResponse execute(HttpRequest request) {
 
         }
 
-        inputStream = conn.getInputStream();
+        try {
```
Comment (mgenov) [inline src/main/java/com/clouway/chita/ChitaHttpClient.java]: I think that this should be placed in the HttpResponse class which constructor need to be protected for the package, to disallow external usages.

Comment (stefan-dimitrov) [inline src/main/java/com/clouway/chita/ChitaHttpClient.java]: Do you mean the getting of the connection attributes (*conn.get...*)?

File: src/main/java/com/clouway/chita/ChitaHttpClient.java [hunk 2]
```diff
@@ -95,7 +95,11 @@ public HttpResponse execute(HttpRequest request) {
...
+          inputStream = conn.getInputStream();
```
Comment (stefan-dimitrov) [inline src/main/java/com/clouway/chita/ChitaHttpClient.java]: Problem from the issue was coming from ` conn.getInputStream() ` which was throwing an IOException when the response from the server contained only an error code.
The exception was being caught [on this line](https://github.com/clouway/chita/pull/13/files#diff-dbe726534ebbf5921a4b3d05830f4671R110) and printed out.
In turn a ` new HttpResponse(-1, ""); ` was being returned([from this line](https://github.com/clouway/chita/pull/13/files#diff-dbe726534ebbf5921a4b3d05830f4671R125))

File: src/main/java/com/clouway/chita/ChitaHttpClient.java (hunk 1 above)
Comment (mgenov) [inline src/main/java/com/clouway/chita/ChitaHttpClient.java]: Yeah, it should be in the Response.

**Comment by stefan-dimitrov** [comment]
Connection instance is being passed to *HttpResponse* constructor as [suggested](https://github.com/clouway/chita/pull/13#discussion_r26032143). I'm not sure I did it properly, please have a look - [code pointer](https://github.com/clouway/chita/pull/13/files#diff-1a319e86953811b6ae48f3f76f7ec0c6R38)

**Comment by mgenov** [comment]
It doesn't look well, cause some work is done in the constructor. I think that we can keep reference to the connection in the response and use it for each of the methods.

**Comment by stefan-dimitrov** [comment]
I'll give it a go.

**Comment by stefan-dimitrov** [comment]
I'm having trouble getting it to work when getting the connection instance attributes in each method.
It seems that the connection instance goes *null* when some of it's getters are called.

Here is what I have for the HttpResponse class:
```java
public class HttpResponse {

  public interface ResponseRead<E> {
    E as(Class<? extends ChitaTransport> var1);
  }

  public static HttpResponse dummyResponse() {
    return new HttpResponse(0, "");
  }

  private int code;

  private  String message;

  private InputStream inputStream;

  private HttpURLConnection connection;


  HttpResponse(int code, String message) {

    this.code = code;
    this.message = message;
  }

  HttpResponse(HttpURLConnection connection) {

    this.connection = connection;
  }

  /**
   * Returns the HTTP status code.
   */
  public int code() {
    if (connection == null) {
      return code;
    }
    try {
      code = connection.getResponseCode();
    } catch (IOException e) {
    }
    return code;
  }

  /**
   * Returns true if the code is in [200..300), which means the request was
   * successfully received, understood, and accepted.
   */
  public boolean isSuccessful() {
    return code() == 200;
  }

  /**
   * the response is dummy when the request was not sent because the
   * given url(destination) was empty or NULL
   *
   * @return
   */
  public boolean isDummy() {
    if (code() == 0) {
      return true;
    }
    return false;
  }

  /**
   * Returns the HTTP status message or null if it is unknown.
   */
  public String statusMessage() {
    if (connection == null) {
      return "null";
    }
    try {
      return connection.getResponseMessage();
    } catch (IOException e) {
      return "null";
    }
  }

  public <E> ResponseRead<E> read(final Class<E> entityClazz) {
    return new ResponseRead<E>() {
      @Override
      public E as(Class<? extends ChitaTransport> clazz) {
        E result = null;
        try {
          ChitaTransport transport = clazz.newInstance();
          result = transport.in(connection.getInputStream(), entityClazz);
        } catch (InstantiationException e) {
          e.printStackTrace();
        } catch (IllegalAccessException e) {
          e.printStackTrace();
        } catch (IOException e) {
          e.printStackTrace();
        }
        return result;
      }
    };
  }

  /**
   * Returns the response as byte array
   */
  public byte[] readBytes(){
    try {
      return IOUtils.toByteArray(connection.getInputStream());
    } catch (IOException e) {
      e.printStackTrace();
    }
    return null;
  }
}
```

**Comment by mgenov** [comment]
I don't get your point, but have you looked at the getErrorStream? 
getErrorStream: 

```
Returns the error stream if the connection failed but the server sent useful data nonetheless. The typical example is when an HTTP server responds with a 404, which will cause a FileNotFoundException to be thrown in connect, but the server sent an HTML help page with suggestions as to what to do. This method will not cause a connection to be initiated. If the connection was not connected, or if the server did not have an error while connecting or if the server had an error but no error data was sent, this method will return null. This is the default.
```


Here is an example snippet that verifies the behaviour: 

```java
    testServer.pretendThatServiceReplaysWithStatus(400);
    testServer.pretendThatNextResponseWillBe("test message");

    URL url = new URL(testServer.getTestEndpoint());

    HttpURLConnection connection = (HttpURLConnection) url.openConnection();

    int code = connection.getResponseCode();
    InputStream stream = null;
    if (connection.getErrorStream() != null) {
      stream = connection.getErrorStream();
    } else {
      stream = connection.getInputStream();
    }
    System.out.println(new String(ByteStreams.toByteArray(stream)));
```

so in case of wrong error code, the content could be retrieved from the error stream which could be checked for nulls. 

You can still pass InputStream to the constructor of the response as and status code.

**Comment by mgenov** [comment]
Note that the following snippet works with status code 200 and 400.

**Comment by stefan-dimitrov** [comment]
Sorry, I wasn't clear. My point is that when I gather the attributes of *connection* in one go there is no problem. But when I gather each one individually, in separate methods of HttpResponse, then I don't get the values I'm supposed to get and many tests fail.

About getErrorStream, yes I used it in first [commit](https://github.com/stefan-dimitrov/chita/commit/7f13544bc36aef54010d9f7bab42433978f93ca1#diff-dbe726534ebbf5921a4b3d05830f4671R101) of this PR. I have forgotten to put it here.

**Comment by mgenov** [comment]
If there is no problem with gathering, then pass inpustream and status code instead of connection.

**Comment by stefan-dimitrov** [comment]
I think it will be easier that way. I'll revert the changes.

**Comment by stefan-dimitrov** [comment]
Reverted.

File: src/main/java/com/clouway/chita/HttpClient.java [hunk 3]
```diff
@@ -94,7 +94,11 @@
 
         }
 
-        inputStream = conn.getInputStream();
+        try {
```
Comment (mgenov) [inline src/main/java/com/clouway/chita/HttpClient.java]: Why you are not using if statement. It will be faster then waiting to throw an exception.

Comment (stefan-dimitrov) [inline src/main/java/com/clouway/chita/HttpClient.java]: I tried using *if*, but apparently there is no error stream and code goes directly to getInputStream and the bug would persist.

Comment (mgenov) [inline src/main/java/com/clouway/chita/HttpClient.java]: Yeah, error stream is available only when response code is not 200. 

If you check error stream for null you will get the correct result I think. 

200 -> conn.getInputStream(); 
Any other code -> conn.getErrorStream();

**Comment by stefan-dimitrov** [comment]
Made the if-else to work for getting of the input stream.

**Comment by mgenov** [comment]
LGTM !

**Pull request merged**